bin/add_frontend_server.rb

```diff
@@ -0,0 +1,120 @@
+
+determined_instance_zone = instance_distribution.select { |_, v| v == instance_distribution.values.min }.keys[0]
+
+puts "Using underscaled instance zone #{determined_instance_zone}"
+
+instance_name = 'frontend-' + determined_instance_zone[-1, 1] + (instance_distribution[determined_instance_zone] + 1).to_s
+puts "Naming instance #{instance_name}"
+
+run_instance_response = ec2client.run_instances ({
+                                                    dry_run: false,
+                                                    min_count: 1,
+                                                    max_count: 1,
+                                                    image_id: 'ami-d05e75b8',  #Image ID for ubuntu instance we use
+                                                    instance_type: 'c3.8xlarge',
+                                                    monitoring: {
+                                                        enabled:true
+                                                    },
+                                                    disable_api_termination: true, #Prevent against accidental termination
+                                                    placement: {
+                                                        availability_zone: determined_instance_zone
+                                                    },
+                                                    block_device_mappings: [
+                                                        {
+                                                            device_name: '/dev/sda1',
+                                                            ebs: {
+                                                                volume_size: 128,
+                                                                delete_on_termination: true,
+                                                                volume_type: 'gp2',
+                                                            },
+                                                        },
+                                                    ],
+                                                    security_groups: ['pegasus'],
+                                                })
+
+instance_id = run_instance_response.instances[0].instance_id
+puts "Looking for instance id  #{instance_id}"
+time_waited = 0
+
+while time_waited < MAX_WAIT_TIME
+  time_waited += 5
+  sleep(5)
+  puts 'Checking for instance creation'
+  instance_status_response = ec2client.describe_instance_status({instance_ids:[instance_id]})
+
+  unless instance_status_response.instance_statuses.empty?
+    instance_state = instance_status_response.instance_statuses[0].instance_state.name
+    puts "Instance ID #{instance_id} is now #{instance_state}"
+    if instance_state == 'running'
+      break
+    end
+  end
+end
+
+unless instance_state == 'running'
+  puts "Unable to create instance after waiting #{time_waited} seconds"
+  exit(1)
+end
```

Comment:
Recent versions of the Ruby AWS SDK provide [`wait_until`](http://docs.aws.amazon.com/sdkforruby/api/Aws/EC2/Client.html#wait_until-instance_method) methods that make this kind of long-polling logic much more concise: `ec2client.wait_until(:instance_status_ok, instance_ids:[instance_id])` should be able to replace lines 82-103 here.